theano/scan_module/scan_opt.py

```diff
@@ -207,6 +209,32 @@ def remove_constants_and_unused_inputs_scan(node):
     else:
         return False
 
+def retrieve_clients_fromnodes(nodes):
+    clients = []
+    if not isinstance(nodes, list):
+        nodes = [nodes]
+
+    for node in nodes:
+        if isinstance(node, theano.Apply):
+            node_outs = node.outputs
+            for node_out in node_outs:
+                for ncli in node_out.clients:
+                    if isinstance(ncli[0], list) or isinstance(ncli[0], tuple):
+                        if ncli[0][0] != "output" and ncli[0][0] != node:
+                            clients.append(ncli[0][0])
+                    else:
+                        if ncli[0] != "output" and ncli[0] != node:
+                           clients.append(ncli[0])
+        else:
+            for ncli in node.clients:
+                if isinstance(ncli[0], list) or isinstance(ncli[0], tuple):
+                    if ncli[0][0] != "output" and ncli[0][0] != node:
+                        clients.append(ncli[0][0])
+                else:
+                    if ncli[0] != "output" and ncli[0] != node:
+                        clients.append(ncli[0])
+
+    return clients
```

Comment:
This function can be made much simpler : 
- A node can never be a client of one of it's own inputs.
- A variable's clients is always a list of tuples. The first element of each tuple will be either 'output' or an apply node
- There is much code duplication; the code can be refactored for readability